### PR TITLE
Add Animal Pose Preprocessor (RTMPose AP-10K)

### LINKED ISSUE #2258: [Feature Request]: Animal Pose Preprocessor (RTMPose AP-10K)
https://github.com/abehonest/ControlNet_AnimalPose
DWPose but for animal
The original implementation doesn't use a object detector so it can only recognize one animal in the image. From my test, YOLOX used by DWPose works pretty well.
MMPose can be replaced with this [onnx checkpoint](https://download.openmmlab.com/mmpose/v1/projects/rtmposev1/onnx_sdk/rtmpose-m_simcc-ap10k_pt-aic-coco_210e-256x256-7a041aa1_20230206.zip)

### PR DESCRIPTION
**Pull request by sdbds**
Closes #2258.

image

I will training a controlnet-lllite for animal pose, too

image
image

I tested openpose and dwpose, they work well.

### CODE REVIEW

**Comment by iTeddyBe** [comment]
Hi @sdbds
check it

onnx_pose  = load_model("rtmpose-m_simcc-ap10k_pt-aic-coco_210e-256x256-7a041aa1_20230206.onnx", animal_onnx)

onnx_pose  = load_model("rtmpose-m_simcc-ap10k_pt-aic-coco_210e-256x256-7a041aa1_20230206.onnx", **animal_onnx_pose**)

**Comment by sdbds** [comment]
> Hi @sdbds check it
> 
> onnx_pose = load_model("rtmpose-m_simcc-ap10k_pt-aic-coco_210e-256x256-7a041aa1_20230206.onnx", animal_onnx)
> 
> onnx_pose = load_model("rtmpose-m_simcc-ap10k_pt-aic-coco_210e-256x256-7a041aa1_20230206.onnx", **animal_onnx_pose**)

Thank you for noticing,i fix it.

File: annotator/openpose/animalpose.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+            bottom=bottom,
+            left=left,
+            right=right,
+            borderType=cv2.BORDER_REPLICATE,
+        )
+
+    return paddedImg
+
+
+def smartCrop(img, size, center):
```
Comment (huchenlei) [inline annotator/openpose/animalpose.py:92]: Does not seem to be referenced.

File: annotator/openpose/animalpose.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+            yCenter -= -upMargin
+        if downMargin > height:
+            yCenter -= downMargin - height
+
+        img = cv2.getRectSubPix(img, size, (xCenter, yCenter))
+
+    return img
+
+
+def calculateScaleFactor(img, size, poseSpanX, poseSpanY):
```
Comment (huchenlei) [inline annotator/openpose/animalpose.py:121]: Unused function.

File: annotator/openpose/animalpose.py [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+    scaleFactorY = 1
+    if poseSpanY > size[1]:
+        scaleFactorY = size[1] / poseSpanY
+
+    scaleFactor = min(scaleFactorX, scaleFactorY)
+
+    return scaleFactor
+
+
+def scaleImg(img, size, poseSpanX, poseSpanY, scaleFactor):
```
Comment (huchenlei) [inline annotator/openpose/animalpose.py:138]: Unused function.

File: annotator/openpose/animalpose.py [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+def drawBetweenSetofKeypointLists(
+    pose_img, keypoints_set, keypointPairsList, colorsList, scaleFactor
+):
+    for keypoints in keypoints_set:
+        drawBetweenKeypointsList(
+            pose_img, keypoints, keypointPairsList, colorsList, scaleFactor
+        )
+
+
+def padImg(img, size, blackBorder=True):
```
Comment (huchenlei) [inline annotator/openpose/animalpose.py:46]: Unused function.

File: annotator/openpose/animalpose.py [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+def drawBetweenKeypointsList(
+    pose_img, keypoints, keypointPairsList, colorsList, scaleFactor
+):
+    for ind, keypointPair in enumerate(keypointPairsList):
+        drawBetweenKeypoints(
+            pose_img, keypoints, keypointPair, colorsList[ind], scaleFactor
+        )
+
+
+def drawBetweenSetofKeypointLists(
```
Comment (huchenlei) [inline annotator/openpose/animalpose.py:37]: Unused function.

**Comment by huchenlei** [comment]
LGTM.

Note:
I flatten the keypoint list in json result so that it is consistent with human openpose json format.

Now the JSON looks like:
```typescript
{
  canvas_height: number;
  canvas_width: number;
  version: string;
  animals: number[][];
}
```

**Comment by sdbds** [comment]
> @sdbds Where do we get the ControlNet model for animal poses? The model offered in https://github.com/abehonest/ControlNet_AnimalPose is 8GB (Full precision + gradient weights). We should probably prune it and upload it to hugginface so that it becomes more accessible.

I'm having some trouble recognizing this model when I convert it successfully.
image

**Comment by huchenlei** [comment]
I saved the fp16 model and uploaded it to huggingface here:
https://huggingface.co/huchenlei/animal_openpose/blob/main/control_openpose-fp16.safetensors

It seems to be working fine. Only 700MB in size.

**Comment by huchenlei** [comment]
I have added support for animal openpose edit. sd-webui-openpose-edit start to support edit of animal openpose from v0.2.4.

demo (2)

**Comment by huchenlei** [comment]
The animal openpose used with human openpose might be useful to draw half human half animal creatures like centaur. 
Centaur
Centaur2
CentaurCtrl

**Comment by huchenlei** [comment]
@sdbds Please take a look and make sure everything working as expected. Thanks!

**Comment by sdbds** [comment]
> @sdbds Please take a look and make sure everything working as expected. Thanks!

I'll test the birdman or angel wing effect.

**Comment by huchenlei** [comment]
@sdbds 
I have rebased the branch on origin/main. So you might get conflict you directly pull from remote. The best way to handle this might be just checkout the PR to a new local branch?

**Comment by sdbds** [comment]
> @sdbds I have rebased the branch on origin/main. So you might get conflict you directly pull from remote. The best way to handle this might be just checkout the PR to a new local branch?

image
Everything seems works well,only 1 bug i found is cant click 4buttons in right when switch to openpose, i gusee it is a UI layout bug.
When i click one of 4buttons,it will be a upload json function.

**Comment by huchenlei** [comment]
> > @sdbds I have rebased the branch on origin/main. So you might get conflict you directly pull from remote. The best way to handle this might be just checkout the PR to a new local branch?
> 
> image Everything seems works well,only 1 bug i found is cant click 4buttons in right when switch to openpose, i gusee it is a UI layout bug. When i click one of 4buttons,it will be a upload json function.

I cannot reproduce this issue locally. Can you attach your reproduction steps?

**Comment by sdbds** [comment]
> > > @sdbds I have rebased the branch on origin/main. So you might get conflict you directly pull from remote. The best way to handle this might be just checkout the PR to a new local branch?
> > 
> > 
> > image Everything seems works well,only 1 bug i found is cant click 4buttons in right when switch to openpose, i gusee it is a UI layout bug. When i click one of 4buttons,it will be a upload json function.
> 
> I cannot reproduce this issue locally. Can you attach your reproduction steps?

image
image
image

reproduction steps:
1、Open controlnet tab
2、Change Control Type to Openpose
3、It will show upload json button
4、Click all 4 button will be upload json

I guess upload json button effect layout.

**Comment by huchenlei** [comment]
Thanks! I think the root cause is https://github.com/Mikubill/sd-webui-controlnet/issues/2332. I am going to fix it now.

**Comment by huchenlei** [comment]
Style issue fixed. I think we can submit this PR and start writing a release note.

**Pull request merged**